Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -2145,37 +2145,60 @@ extension BrowserViewController: WKUIDelegate {
-            // since not calling it will result in a runtime exception.
+        var messageAlert = MessageAlert(message: message, frame: frame, completionHandler: completionHandler, suppressHandler: nil)
+        handleAlert(webView: webView, alert: &messageAlert) { (val) in
             completionHandler()
         }
     }
 
     func webView(_ webView: WKWebView, runJavaScriptConfirmPanelWithMessage message: String, initiatedByFrame frame: WKFrameInfo, completionHandler: @escaping (Bool) -> Void) {
-        let confirmAlert = ConfirmPanelAlert(message: message, frame: frame, completionHandler: completionHandler)
-        if shouldDisplayJSAlertForWebView(webView) {
-            present(confirmAlert.alertController(), animated: true, completion: nil)
-        } else if let promptingTab = tabManager[webView] {
-            promptingTab.queueJavascriptAlertPrompt(confirmAlert)
-        } else {
-            completionHandler(false)
+        var confirmAlert = ConfirmPanelAlert(message: message, frame: frame, completionHandler: completionHandler, suppressHandler: nil)
+        handleAlert(webView: webView, alert: &confirmAlert) { (val) in
+            completionHandler(val as? Bool ?? false)
         }
     }
 
     func webView(_ webView: WKWebView, runJavaScriptTextInputPanelWithPrompt prompt: String, defaultText: String?, initiatedByFrame frame: WKFrameInfo, completionHandler: @escaping (String?) -> Void) {
-        let textInputAlert = TextInputAlert(message: prompt, frame: frame, completionHandler: completionHandler, defaultText: defaultText)
+        var textInputAlert = TextInputAlert(message: prompt, frame: frame, completionHandler: completionHandler, defaultText: defaultText, suppressHandler: nil)
+        handleAlert(webView: webView, alert: &textInputAlert) { (val) in
+            completionHandler(val as? String)
+        }
+    }
+    
+    func handleAlert<T: JSAlertInfo>(webView: WKWebView, alert: inout T, completionHandler: @escaping (Any?) -> Void) {
+        guard let promptingTab = tabManager[webView], !promptingTab.blockAllAlerts else {
+            tabManager[webView]?.cancelQueuedAlerts()
+            completionHandler(nil)
+            return
+        }
+        promptingTab.alertShownCount += 1
+        var suppressBlock: JSAlertInfo.SuppressHandler = ({[unowned self] suppress in
+            if suppress {
+                func suppressDialogues() {
+                    promptingTab.blockAllAlerts = true
+                    self.tabManager[webView]?.cancelQueuedAlerts()
+                    completionHandler(nil)
+                }
+                // Show confirm alert here.
+                let suppressSheet = UIAlertController(title: nil, message: Strings.SuppressAlertsActionMessage, preferredStyle: .actionSheet)
+                suppressSheet.addAction(UIAlertAction(title: Strings.SuppressAlertsActionTitle, style: .destructive, handler: { (action) in
```

Comment:
Probably cleaner to make `suppressDialogues` take an empty `UIAlertAction`:

`func suppressDialogues(_: UIAlertAction) {`

Since then you can just pass it in without creating a new closure around it:

```
suppressSheet.addAction(UIAlertAction(title: Strings.SuppressAlertsActionTitle, style: .destructive, handler: suppressDialogues))
```